linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java

```diff
@@ -0,0 +1,215 @@
+  }
+
+  static synchronized Analytics getInstance(boolean sendAnalytics) {
+    if (instance == null) {
+      String collectionUrl = null;
+      if (!PlaceholderConstants.FIRELOG_API_KEY.startsWith("@")) {
+        collectionUrl = FIRELOG_COLLECTION_URL + "?key=" + PlaceholderConstants.FIRELOG_API_KEY;
+      }
+      instance = new Analytics(collectionUrl, new ConcurrentLinkedQueue<>(), sendAnalytics);
+      instance.start();
+    }
+    return instance;
+  }
+
+  /**
+   * Sends a usage metric to Google Analytics.
+   */
+  void queuePing(String eventName, String metadataKey, String metadataValue) {
+    Preconditions.checkNotNull(metadataKey, "metadataKey null");
+    Preconditions.checkArgument(!metadataKey.isEmpty(), "metadataKey empty");
+    Preconditions.checkNotNull(metadataValue, "metadataValue null");
+    Preconditions.checkArgument(!metadataValue.isEmpty(), "metadataValue empty");
+
+    queuePing(eventName, ImmutableMap.of(metadataKey, metadataValue));
+  }
+
+  /**
+   * Sends a usage metric to Google Analytics.
+   */
+  void queuePing(String eventName, Map<String, String> metadata) {
+    Preconditions.checkNotNull(eventName, "eventName null");
+    Preconditions.checkArgument(!eventName.isEmpty(), "eventName empty");
+    Preconditions.checkNotNull(metadata);
+
+    if (collectionUrl != null) {
+      ImmutableMap<String, String> metadataCopy = ImmutableMap.copyOf(metadata);
+      pingEventQueue.add(new PingEvent(eventName, metadataCopy));
+    }
+  }
+  
+  void queuePing(String eventName) {
+    queuePing(eventName, ImmutableMap.<String, String>of());
+  }
+
+  /**
+   * This is the only method that makes an HTTP connection. Everything else
+   * ultimately funnels through here. 
+   */
+  private void ping(PingEvent pingEvent) {
+    if (sendAnalytics && pingEvent != null) {
+      try {
+        String json = jsonEncode(pingEvent);
+        int resultCode = HttpUtil.sendPost(collectionUrl, json, "application/json");
+        if (resultCode >= 300) {
+          logger.log(Level.FINE, "Failed to POST to Concord with HTTP return code " + resultCode);
+        }
+      } catch (IOException ex) {
+        // Don't recover or retry.
+        logger.log(Level.FINE, "Failed to POST to Concord", ex);
+      } 
+    }
+  }
+
+  @VisibleForTesting
+  static class PingEvent {
+    @VisibleForTesting final String eventName;
+    @VisibleForTesting final ImmutableMap<String, String> metadata;
+
+    PingEvent(String eventName, ImmutableMap<String, String> metadata) {
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(eventName), "eventName null or empty");
+      Preconditions.checkNotNull(metadata, "metadata is null");
+      this.eventName = eventName;
+      this.metadata = metadata;
+    }
+  }
+
+  @VisibleForTesting
+  String jsonEncode(PingEvent event) {
+    Gson gson = new Gson();
+
+    Map<String, String> desktopClientInfo = new HashMap<>();
+    desktopClientInfo.put("os", System.getProperty("os.name"));
```

Comment:
Linkage Monitor does not target for desktops.

Reply:
I think this is required for logging though. @chanseokoh Do you recall?

Reply:
I meant the variable name.

```suggestion
    Map<String, String> osInfo = new HashMap<>();
    osInfo.put("os", System.getProperty("os.name"));
```

I don't have any concern on sending operating system names but I expect they are all Linux.

Reply:
on line 184 note that this value is stored in a field with the name "desktop_client_info" as defined by the proto